### PR TITLE
feat: implement generator and container interfaces

### PR DESCRIPTION
**Pull request by julieqiu**
Create initial implementation of generator and container interfaces.

### CODE REVIEW

File: internal/command/command.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		if _, err := cloneGoogleapis(ctx); err != nil {
+			return err
+		}
+		return container.Generate(ctx, flagLanguage, flagAPIRoot, flagAPIPath, flagOutput, flagGeneratorInput)
+	},
+}
+
+var CmdUpdateRepo = &Command{
+	Name:  "update-repo",
+	Short: "Configure a new API in a given language",
```
Comment (jskeet) [inline internal/command/command.go:92]:
```suggestion
	Short: "Update a language repository with new API definitions",
```

(Or similar...)

File: internal/command/command.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		if flagAPIRoot == "" {
+			return fmt.Errorf("-api-root is not provided")
+		}
+		if flagAPIPath == "" {
+			return fmt.Errorf("-api-path is not provided")
+		}
+		if !supportedLanguages[flagLanguage] {
+			return fmt.Errorf("invalid -language flag specified: %q", flagLanguage)
+		}
+		if _, err := cloneGoogleapis(ctx); err != nil {
```
Comment (jskeet) [inline internal/command/command.go:83]: We only need to clone googleapis if api-root isn't specified (in which case, the location of the clone is the default)

File: internal/command/command.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		addFlagOutput,
+	} {
+		fn(fs)
+	}
+
+	fs = CmdUpdateRepo.flags
+	for _, fn := range []func(fs *flag.FlagSet){
+		addFlagAPIPath,
+		addFlagAPIRoot,
+		addFlagBranch,
```
Comment (jskeet) [inline internal/command/command.go:171]: Not used yet?

File: internal/container/container.go [hunk 4]
```diff
@@ -14,22 +14,53 @@
...
+func runGenerate(googleapisDir, languageDir, api string) error {
+	args := []string{
+		"run",
+		"-v", fmt.Sprintf("%s:/apis", googleapisDir),
+		"-v", fmt.Sprintf("%s:/output", languageDir),
+		dotnetImageTag,
+		"--command=update",
+		"--api-root=/apis",
+		fmt.Sprintf("--api=%s", api),
+		"--output-root=/output",
```
Comment (jskeet) [inline internal/container/container.go:53]: Just --output now

File: internal/container/container.go [hunk 5]
```diff
@@ -14,22 +14,53 @@
...
-	return errNotImplemented
+func Configure(ctx context.Context, language, apiRoot, apiPath, generatorInput string) error {
+	return runCommand("echo", "configure not implemented")
 }
 
-func Build() error {
-	return errNotImplemented
+const dotnetImageTag = "picard"
+
+func runGenerate(googleapisDir, languageDir, api string) error {
```
Comment (jskeet) [inline internal/container/container.go:44]: Let's change "api" to "apiPath" to be clear

File: internal/container/container.go [hunk 6]
```diff
@@ -14,22 +14,53 @@
...
+	args := []string{
+		"run",
+		"-v", fmt.Sprintf("%s:/apis", googleapisDir),
+		"-v", fmt.Sprintf("%s:/output", languageDir),
+		dotnetImageTag,
+		"--command=update",
+		"--api-root=/apis",
+		fmt.Sprintf("--api=%s", api),
```
Comment (jskeet) [inline internal/container/container.go:52]: This should be api-path, but only if it's actually specified.
(Or we could make the container contract say that it's *always* specified, and to ignore it if it's empty? I don't much mind either way.)

File: internal/container/container.go [hunk 7]
```diff
@@ -14,22 +14,53 @@
...
 
-func Clean() error {
-	return errNotImplemented
+func Configure(ctx context.Context, language, apiRoot, apiPath, generatorInput string) error {
+	return runCommand("echo", "configure not implemented")
 }
 
-func Build() error {
-	return errNotImplemented
+const dotnetImageTag = "picard"
```
Comment (jskeet) [inline internal/container/container.go:42]: This is the bit I'll fix soon, with any luck.

**Review by jskeet — APPROVED** [review]
Various tweaks needed, but we could merge and then I can work on the fixes if you'd prefer? Approving for that reason.

File: internal/command/command.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		}
+		return container.Generate(ctx, flagLanguage, flagAPIRoot, flagAPIPath, flagOutput, flagGeneratorInput)
+	},
+}
+
+var CmdUpdateRepo = &Command{
+	Name:  "update-repo",
+	Short: "Configure a new API in a given language",
+	Run: func(ctx context.Context) error {
+		if flagAPIRoot == "" {
```
[the remark this answers is not in the document]
Comment (julieqiu) [reply internal/command/command.go:94]: yup going to resolve this and add that in a future PR

File: internal/command/command.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+		if !supportedLanguages[flagLanguage] {
+			return fmt.Errorf("invalid -language flag specified: %q", flagLanguage)
+		}
+		if _, err := cloneGoogleapis(ctx); err != nil {
+			return err
+		}
+		if _, err := cloneLanguageRepo(ctx, flagLanguage); err != nil {
+			return err
+		}
+		if err := container.Generate(ctx, flagLanguage, flagAPIRoot, flagAPIPath, flagOutput, flagGeneratorInput); err != nil {
```
[the remark this answers is not in the document]
Comment (julieqiu) [reply internal/command/command.go:109]: Filed https://github.com/googleapis/generator/issues/26

File: internal/container/container.go (hunk 6 above)
Comment (julieqiu) [reply internal/container/container.go:52]: I went with the second for now. We can change it later.

File: internal/command/command.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+			return err
+		}
+		if err := container.Clean(ctx, flagLanguage, flagOutput, flagAPIPath); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (julieqiu) [reply internal/command/command.go:114]: Filed https://github.com/googleapis/generator/issues/27

File: internal/command/command.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (julieqiu) [reply internal/command/command.go:108]: I'm still not sure I understand what this means, but I'll merge and leave it to you to address.

**Pull request merged**